### ISSUE TITLE
OpenMP: Class method call results in segfault

### ISSUE THREAD

**Issue by pawosm-arm** [flang-compiler/flang#347]
Having following code:
```
  SUBROUTINE FOO(C)
    IMPLICIT NONE
    CLASS(CIRCLE) :: C
    INTEGER :: I

    !$OMP PARALLEL DO SCHEDULE(DYNAMIC) PRIVATE(I)
    DO I=1,10
      CALL C%PRINT_RADIUS()
    ENDDO
    !$OMP END PARALLEL DO
  END SUBROUTINE FOO
```
With flang1 patched to fully capture F2003 class objects into outlined parallel region, I'm observing segfault at method call site in outlined parallel region.

Quick examination of LLVM-IR shows slight change in the affected generated code before and after outlining, before:
```
%0 = bitcast i64* %.O0000 to i8*
%1 = getelementptr i8, i8* %0, i64 40
```
after:
```
%39 = bitcast i8* %38 to i8***
%40 = load i8**, i8*** %39, align 8
%41 = load i8*, i8** %40, align 8
%42 = getelementptr i8, i8* %41, i64 40
```

These extra load instructions results in one extra 'ld' opcode in the final assembly:
```
0x0000000000401014 <+272>:   ldr     x9, [x9]
0x0000000000401018 <+276>:   ldr     x9, [x9,#40]
```
Instead of loading address at offset 40, a value is read by the first load at offset 0. Since it is rubbish, following use of value in x9 to read value at offset 40 (the second ld) results in segfault.

I managed to run affected binary in gdb, set breakpoint at 0x0000000000401014 and when it was hit, skip to 0x0000000000401018 and as expected, the right method was called on the object used in this parallel region.

**Comment by pawosm-arm** [comment]
Got something!
I tried to break it even more to see if I can come up with some clues how to fix this issue. And I did this:
```
diff --git a/tools/flang2/flang2exe/exp_rte.c b/tools/flang2/flang2exe/exp_rte.c
index 5cdf364..cd8fcf3 100644
--- a/tools/flang2/flang2exe/exp_rte.c
+++ b/tools/flang2/flang2exe/exp_rte.c
@@ -1125,7 +1125,7 @@ exp_type_bound_proc_call(int arg, int descno, int vtoff, int arglnk)
     ili = ad2ili(IL_LDA, ili, NME_UNK);
     ili = ad3ili(IL_AADD, ili, ad_aconi(func_offset), 0);
     ili = ad2ili(IL_LDA, ili, NME_UNK);
-  } else if (SCG(sym) != SC_DUMMY) {
+  } else if (SCG(sym) != SC_DUMMY || 1) {
     ili = mk_address(sym);
     ili = ad3ili(IL_AADD, ili, ad_aconi(type_offset), 0);
     ili = ad2ili(IL_LDA, ili, NME_UNK);
```
...compiler changed that way produced runnable code when started with -fopenmp!
THIS IS NOT A FIX YET though, I suspect that it could break other things.

**Comment by gklimowicz** [comment]
Should be fixed in PR #452 commit 570d7fe
